src/components/PriorityModeController.tsx

```diff
@@ -0,0 +1,97 @@
+    const [isLoadingReportData] = useOnyx(ONYXKEYS.IS_LOADING_REPORT_DATA, {canBeMissing: true});
+    const [isInFocusMode] = useOnyx(ONYXKEYS.NVP_PRIORITY_MODE, {selector: (priorityMode) => priorityMode === CONST.PRIORITY_MODE.GSD, canBeMissing: true});
+    const [hasTriedFocusMode] = useOnyx(ONYXKEYS.NVP_TRY_FOCUS_MODE, {canBeMissing: true});
+    const [allReports] = useOnyx(ONYXKEYS.COLLECTION.REPORT, {canBeMissing: true});
+    const currentRouteName = useCurrentRouteName();
+    const [shouldShowModal, setShouldShowModal] = useState(false);
+    const closeModal = useCallback(() => setShouldShowModal(false), []);
+
+    const validReportCount = useMemo(() => {
+        let count = 0;
+        Object.values(allReports ?? {}).forEach((report) => {
+            if (!isValidReport(report) || !isReportParticipant(accountID ?? CONST.DEFAULT_NUMBER_ID, report)) {
+                return;
+            }
+
+            count++;
+        });
+        return count;
+    }, [accountID, allReports]);
+
+    // We set this when we have finally auto-switched the user of #focus mode to prevent duplication.
+    const hasSwitched = useRef(false);
+
+    // Listen for state changes and trigger the #focus mode when appropriate
+    useEffect(() => {
+        // Wait for Onyx state to fully load
+        if (isLoadingReportData !== false || isInFocusMode === undefined || hasTriedFocusMode === undefined || !accountID) {
+            return;
+        }
+
+        if (hasSwitched.current || isInFocusMode || hasTriedFocusMode) {
+            return;
+        }
+
+        if (validReportCount < CONST.REPORT.MAX_COUNT_BEFORE_FOCUS_UPDATE) {
+            Log.info('[PriorityModeController] Not switching user to focus mode as they do not have enough reports', false, {validReportCount});
+            return;
+        }
+
+        // We wait for the user to navigate back to the home screen before triggering this switch
+        const isNarrowLayout = getIsNarrowLayout();
+        if ((isNarrowLayout && currentRouteName !== SCREENS.HOME) || (!isNarrowLayout && currentRouteName !== SCREENS.REPORT)) {
+            Log.info("[PriorityModeController] Not switching user to focus mode as they aren't on the home screen", false, {validReportCount, currentRouteName});
+            return;
+        }
+
+        Log.info('[PriorityModeController] Switching user to focus mode', false, {validReportCount, hasTriedFocusMode, isInFocusMode, currentRouteName});
+        updateChatPriorityMode(CONST.PRIORITY_MODE.GSD, true);
+        setShouldShowModal(true);
+        hasSwitched.current = true;
+    }, [accountID, currentRouteName, hasTriedFocusMode, isInFocusMode, isLoadingReportData, validReportCount]);
+
+    return shouldShowModal ? <FocusModeNotification onClose={closeModal} /> : null;
+}
+
+/**
+ * A funky but reliable way to subscribe to screen changes.
+ */
+function useCurrentRouteName() {
```

Comment:
Should we put this somewhere that is more likely to be discovered and used like Navigation related files?

Reply:
It's bit hacky so I'm not sure it should be discovered 😄. I think this is the only place we subscribe to screen changes in this way so IMO it's fine to just leave it here for now

Reply:
Ideally we would use [useNavigationState](https://reactnavigation.org/docs/use-navigation-state) for this, but when I tried it, it didn't quite work right 🤷. The current route it pointed to wasn't updating as expected. So I just reimplemented what we were doing before here - subscribe to state changes but then using the navigationRef to get the current route.